File: frontend/src/services/callyzer.ts
import { apiClient } from "./http";


export interface CallyzerConnection {
  id: string;
  label: string | null;
  status: string; // ok | needs_reauth | error
  status_detail: string | null;
  last_synced_at: string | null;
  last_call_at: string | null;
  is_active: boolean;
  created_at: string;
}

/** A call synced from Callyzer (matched to a lead by number at read time). */
export interface ExternalCall {
  id: string;
  provider: string;
  external_id: string;
  client_number: string | null;
  client_name: string | null;
  emp_number: string | null;
  emp_name: string | null;
  call_type: string | null;   // Incoming | Outgoing | Missed | Rejected
  call_method: string | null; // PhoneCall | WhatsAppCall
  call_mode: string | null;   // Voice | Video
  duration_seconds: number | null;
  call_at: string | null;
  note: string | null;
  crm_status: string | null;
  reminder_at: string | null;
  recording_url: string | null;
  created_at: string;
  lead: { id: string; lead_number: number; contact_name: string } | null;
}

export interface CallsListResponse {
  items: ExternalCall[];
  pagination: { page: number; page_size: number; total: number; total_pages: number };
}

export interface CallsQuery {
  call_type?: string;
  q?: string;
  matched?: boolean;
  date_from?: string;
  date_to?: string;
  page?: number;
  page_size?: number;
}

/** Per-employee call performance (Calls → Performance tab). */
export interface EmployeeCallStats {
  user_id: string | null;
  name: string;
  emp_number: string | null;
  unmatched: boolean;
  total: number;
  outgoing: number;
  incoming: number;
  missed: number;
  rejected: number;
  connected: number;
  connect_rate: number; // 0..1
  total_talk_seconds: number;
  avg_talk_seconds: number;
  unique_clients: number;
  last_call_at: string | null;
}

export interface CallStatsTotals {
  callers: number;
  total: number;
  connected: number;
  missed: number;
  total_talk_seconds: number;
  unique_clients: number;
}

export interface CallStatsResponse {
  date_from: string;
  date_to: string;
  totals: CallStatsTotals;
  rows: EmployeeCallStats[];
}

export interface CallStatsQuery {
  date_from?: string;
  date_to?: string;
}

/** A manual caller → FlexCRM-user mapping (fixes unmatched attribution). */
export interface CallAgentMapping {
  id: string;
  emp_key: string;
  emp_number: string | null;
  emp_name: string | null;
  user_id: string;
  user_name: string | null;
  created_at: string;
}

export interface CallTrendPoint {
  date: string; // YYYY-MM-DD (IST day)
  calls: number;
  connected: number;
  missed: number;
}

export interface CallTrendResponse {
  user_id: string;
  date_from: string;
  date_to: string;
  points: CallTrendPoint[];
}


/** Tenant-facing Callyzer connection management (Integrations page). */
export const callyzerService = {
  async list(): Promise<CallyzerConnection[]> {
    const { data } = await apiClient.get<CallyzerConnection[]>("/integrations/callyzer");
    return data;
  },
  async connect(token: string, label: string | null): Promise<CallyzerConnection> {
    const { data } = await apiClient.post<CallyzerConnection>("/integrations/callyzer/connect", {
      token,
      label,
    });
    return data;
  },
  async disconnect(id: string): Promise<void> {
    await apiClient.delete(`/integrations/callyzer/${id}`);
  },
};


/** Reading synced calls — the Calls page + the per-lead list in the lead drawer. */
export const callsService = {
  async list(params: CallsQuery = {}): Promise<CallsListResponse> {
    const { data } = await apiClient.get<CallsListResponse>("/calls", { params });
    return data;
  },
  async forLead(leadId: string): Promise<ExternalCall[]> {
    const { data } = await apiClient.get<ExternalCall[]>(`/calls/lead/${leadId}`);
    return data;
  },
  async stats(params: CallStatsQuery = {}): Promise<CallStatsResponse> {
    const { data } = await apiClient.get<CallStatsResponse>("/calls/stats", { params });
    return data;
  },
  async trend(userId: string, params: CallStatsQuery = {}): Promise<CallTrendResponse> {
    const { data } = await apiClient.get<CallTrendResponse>("/calls/stats/trend", {
      params: { user_id: userId, ...params },
    });
    return data;
  },
  async agents(): Promise<CallAgentMapping[]> {
    const { data } = await apiClient.get<CallAgentMapping[]>("/calls/agents");
    return data;
  },
  async assignAgent(payload: { emp_number: string; emp_name?: string | null; user_id: string }): Promise<CallAgentMapping> {
    const { data } = await apiClient.post<CallAgentMapping>("/calls/agents", payload);
    return data;
  },
  async clearAgent(empKey: string): Promise<void> {
    await apiClient.delete(`/calls/agents/${empKey}`);
  },
};

File: backend/app/schemas/callyzer.py
"""Schemas for the Callyzer call-tracking integration."""
from datetime import date, datetime
from uuid import UUID

from pydantic import Field

from app.schemas.common import ORMModel


# --- Connection (tenant-facing) -------------------------------------------

class CallyzerConnectRequest(ORMModel):
    token: str = Field(min_length=10, max_length=4000)
    label: str | None = Field(default=None, max_length=120)


class CallyzerConnectionRead(ORMModel):
    id: UUID
    label: str | None = None
    status: str = "ok"
    status_detail: str | None = None
    last_synced_at: datetime | None = None
    last_call_at: datetime | None = None
    is_active: bool = True
    created_at: datetime
    # NB: the token is NEVER returned.


# --- Synced call records ---------------------------------------------------

class CallLeadMini(ORMModel):
    id: UUID
    lead_number: int
    contact_name: str


class ExternalCallRead(ORMModel):
    id: UUID
    provider: str = "callyzer"
    external_id: str
    client_number: str | None = None
    client_name: str | None = None
    emp_number: str | None = None
    emp_name: str | None = None
    call_type: str | None = None
    call_method: str | None = None
    call_mode: str | None = None
    duration_seconds: int | None = None
    call_at: datetime | None = None
    note: str | None = None
    crm_status: str | None = None
    reminder_at: datetime | None = None
    recording_url: str | None = None
    created_at: datetime
    # Resolved in the list endpoint (the most-recent lead whose number matches).
    lead: CallLeadMini | None = None


# --- Per-employee call stats (Calls → Performance tab + scorecard factor) ---

class EmployeeCallStats(ORMModel):
    user_id: UUID | None = None          # matched FlexCRM user, if any
    name: str                            # the user's name, else emp_name / emp_number
    emp_number: str | None = None
    unmatched: bool = False              # no FlexCRM user matched this device number
    total: int = 0
    outgoing: int = 0
    incoming: int = 0
    missed: int = 0
    rejected: int = 0
    connected: int = 0                   # calls with talk time > 0
    connect_rate: float = 0.0            # connected / total
    total_talk_seconds: int = 0
    avg_talk_seconds: int = 0            # averaged over connected calls
    unique_clients: int = 0
    last_call_at: datetime | None = None


class CallStatsTotals(ORMModel):
    callers: int = 0
    total: int = 0
    connected: int = 0
    missed: int = 0
    total_talk_seconds: int = 0
    unique_clients: int = 0


class CallStatsResponse(ORMModel):
    date_from: datetime
    date_to: datetime
    totals: CallStatsTotals
    rows: list[EmployeeCallStats]


# --- Caller → user mapping (fix unmatched attribution) ---------------------

class CallAgentAssignRequest(ORMModel):
    emp_number: str = Field(min_length=1, max_length=32)
    emp_name: str | None = Field(default=None, max_length=160)
    user_id: UUID


class CallAgentMappingRead(ORMModel):
    id: UUID
    emp_key: str
    emp_number: str | None = None
    emp_name: str | None = None
    user_id: UUID
    user_name: str | None = None  # resolved in the endpoint
    created_at: datetime


# --- Per-employee daily call trend (Performance drill-down) -----------------

class CallTrendPoint(ORMModel):
    date: date
    calls: int = 0
    connected: int = 0
    missed: int = 0


class CallTrendResponse(ORMModel):
    user_id: UUID
    date_from: datetime
    date_to: datetime
    points: list[CallTrendPoint]
